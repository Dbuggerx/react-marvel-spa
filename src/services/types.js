// @flow

type ResourceItem = {
    resourceURI: string,
    name: string
}

type ResourceItemsWithType = ResourceItem & {
    type: string,
}

type Resources<T> = {
    available: number,
    returned: number,
    collectionURI: string,
    items: T[],
}

export type Hero = {
    id: number,
    name: string,
    description: string,
    modified: string,
    resourceURI: string,
    urls: {
        type: string,
        url: string,
    }[],
    thumbnail: {
        path: string,
        extension: string,
    },
    comics: Resources<ResourceItem>,
    stories: Resources<ResourceItemsWithType>,
    events: Resources<ResourceItem>,
    series: Resources<ResourceItem>,
};

export type HeroesResult = {
    code: number,
    status: string,
    copyright: string,
    attributionText: string,
    attributionHTML: string,
    data: {
        offset: number,
        limit: number,
        total: number,
        count: number,
        results: Hero[],
    },
    etag: string,
};
