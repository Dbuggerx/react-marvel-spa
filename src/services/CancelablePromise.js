// @flow

/**
 * A promise wrapper to handle cancelation
 * @see https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html
 */
export default class CancelablePromise<T> {
    constructor(promise: Promise<T>) {
        this.promise = new Promise((resolve, reject) => {
            promise.then(
                (val: T) => (this.canceled ? reject(new Error(this.constructor.errorMessage)) : resolve(val)),
                (error: Error) => (this.canceled ? reject(new Error(this.constructor.errorMessage)) : reject(error)),
            );
        });
    }
    promise: Promise<T>;
    canceled = false;

    cancel() {
        this.canceled = true;
    }

    static errorMessage = 'canceled';
}
