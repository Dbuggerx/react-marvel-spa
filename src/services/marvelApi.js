// @flow
import type { HeroesResult, Hero } from './types';

const apiKey = '570b1cdca67e5243b4db4425b1b4ee4e';
const imageSizes = {
    small: 'standard_fantastic',
    big: 'portrait_uncanny',
};

/**
 * Fetches data and returns the extracted JSON from the response
 */
async function getJsonFromUrl(url: string) {
    const response = await fetch(url);
    if (!response.ok) {
        throw new Error(response.statusText);
    }
    return response.json();
}

export async function fetchHeros(params: {
    limit: number,
    offset?: number,
    nameStartsWith?: string,
}): Promise<HeroesResult> {
    const queryString = Object.keys(params)
        .map(k => [k, params[k]])
        .reduce((agg, cur) => (cur[1] ? `${agg + (agg.length > 0 ? '&' : '')}${cur[0]}=${String(cur[1])}` : agg), '');

    const url = `https://gateway.marvel.com/v1/public/characters?${queryString}&apikey=${apiKey}`;
    return getJsonFromUrl(url);
}

export async function fetchHeroById(id: number): Promise<HeroesResult> {
    const url = `https://gateway.marvel.com/v1/public/characters/${id}?apikey=${apiKey}`;
    return getJsonFromUrl(url);
}

export function getImageUrl(hero: Hero, size: $Keys<typeof imageSizes>) {
    return `${hero.thumbnail.path}/${imageSizes[size]}.${hero.thumbnail.extension}?apikey=${apiKey}`.replace(
        'http:',
        '',
    );
}
