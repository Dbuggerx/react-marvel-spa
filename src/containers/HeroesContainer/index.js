// @flow
import React, { Component } from 'react';
import { fetchHeros, getImageUrl } from '../../services/marvelApi';
import MainLayout from '../../components/MainLayout';
import ErrorMessage from '../../components/ErrorMessage';
import LoadingIndicator from '../../components/LoadingIndicator';
import HeroList from '../../components/HeroList';
import HeroDetails from '../../components/HeroDetails';
import Information from '../../components/Information';
import Modal from '../../components/Modal';
import PaginationContainer from '../PaginationContainer';
import SearchContainer from '../SearchContainer';
import CancelablePromise from '../../services/CancelablePromise';
import type { Hero } from '../../services/types';

type Props = {
    pageSize: number,
};

type State = {
    isLoading: boolean,
    heroes: Hero[],
    error?: string,
    selection: ?{
        hero: Hero,
        imageUrl: string,
    },
    pagination: ?{
        offset: number,
        limit: number,
        total: number,
        count: number,
    },
    query: ?{
        search?: string,
        offset: number,
    },
    attribution: string,
};

/**
 * Handles the data for the Heroes and passes props to children components
 */
export default class HeroListContainer extends Component<Props, State> {
    constructor(props: Props) {
        super(props);
        this.state = {
            heroes: [],
            isLoading: true,
            selection: null,
            pagination: null,
            query: null,
            attribution: 'Data provided by Marvel. Â© 2014 Marvel',
        };
    }

    componentDidMount() {
        this.fetchMarvelHeroes();
    }

    componentDidUpdate(prevProps: Props, prevState: State) {
        if (this.state.query && this.state.query !== prevState.query) {
            this.fetchMarvelHeroes(this.state.query.offset, this.state.query.search || '');
        }
    }

    componentWillUnmount() {
        if (this.cancelableFetch) {
            this.cancelableFetch.cancel();
        }
    }

    getThumbnailUrl = (hero: Hero) => getImageUrl(hero, 'small');

    cancelableFetch: *;

    handleHeroClick = (hero: Hero) => {
        this.setState({
            selection: {
                hero,
                imageUrl: getImageUrl(hero, 'big'),
            },
        });
    };

    handleCloseModal = () => {
        this.setState({
            selection: null,
        });
    };

    handleSearchChanged = (param: string) => {
        this.setState({
            query: {
                search: param,
                offset: 0,
            },
        });
    };

    fetchHeroesPage = (page: number) => {
        this.setState({
            query: {
                ...this.state.query,
                offset: this.props.pageSize * (page - 1),
            },
        });
    };

    async fetchMarvelHeroes(offset: number = 0, nameStartsWith: string = '') {
        this.setState({
            isLoading: true,
        });

        try {
            // Wrapping the result promise in a "CancelablePromise", in order to be safe to unmount
            this.cancelableFetch = new CancelablePromise(fetchHeros({
                limit: this.props.pageSize,
                offset,
                nameStartsWith,
            }));
            const heroesResult = await this.cancelableFetch.promise;
            this.setState({
                heroes: heroesResult.data.results,
                isLoading: false,
                pagination: {
                    offset: heroesResult.data.offset,
                    limit: heroesResult.data.limit,
                    total: heroesResult.data.total,
                    count: heroesResult.data.count,
                },
                attribution: heroesResult.attributionText,
            });
        } catch (error) {
            if (error.message !== CancelablePromise.errorMessage) {
                this.setState({
                    error: 'Error loading heroes',
                });
            }
        }
    }

    render() {
        if (this.state.error) {
            return <ErrorMessage message={this.state.error} />;
        }
        if (this.state.isLoading) {
            return <LoadingIndicator />;
        }
        return (
            <React.Fragment>
                {this.state.selection && (
                    <Modal title={this.state.selection.hero.name} onClose={this.handleCloseModal}>
                        <HeroDetails hero={this.state.selection.hero} imageUrl={this.state.selection.imageUrl} />
                    </Modal>
                )}
                <MainLayout
                    search={
                        <SearchContainer
                            lastSearch={this.state.query ? this.state.query.search || '' : ''}
                            onSearchChanged={this.handleSearchChanged}
                        />
                    }
                    list={
                        <HeroList
                            onHeroClick={this.handleHeroClick}
                            heroes={this.state.heroes}
                            getThumbnailUrl={this.getThumbnailUrl}
                        />
                    }
                    pagination={
                        this.state.heroes.length > 0 &&
                        this.state.pagination && (
                            <PaginationContainer {...this.state.pagination} goToPage={this.fetchHeroesPage} />
                        )
                    }
                    info={<Information attribution={this.state.attribution} />}
                />
            </React.Fragment>
        );
    }
}
