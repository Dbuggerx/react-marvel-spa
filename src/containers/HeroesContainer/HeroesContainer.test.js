import React from 'react';
import { shallow } from 'enzyme';
import HeroesContainer from '.';
import { fetchHeros, getImageUrl } from '../../services/marvelApi';

jest.mock('../../services/marvelApi');
const successfulResultMock = {
    code: 200,
    status: 'Ok',
    data: {
        offset: 111,
        limit: 222,
        total: 333,
        count: 444,
        results: [
            {
                id: 1011334,
                name: '3-D Man',
                description: '',
                modified: '2014-04-29T14:18:17-0400',
                thumbnail: {
                    path: 'http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784',
                    extension: 'jpg',
                },
            },
            {
                id: 1017100,
                name: 'A-Bomb (HAS)',
                description:
                    "Rick Jones has been Hulk's best bud since day one, but now he's more than a friend...he's a teammate! Transformed by a Gamma energy explosion, A-Bomb's thick, armored skin is just as strong and powerful as it is blue. And when he curls into action, he uses it like a giant bowling ball of destruction! ",
                modified: '2013-09-18T15:54:04-0400',
                thumbnail: {
                    path: 'http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16',
                    extension: 'jpg',
                },
            },
        ],
    },
};

describe('HeroesContainer', () => {
    beforeEach(() => {
        fetchHeros.mockClear();
        getImageUrl.mockClear();
    });

    describe('render', () => {
        test('renders a loading message', () => {
            // A simple mocked Promise will do to test the loading message
            fetchHeros.mockReturnValue(Promise.resolve());

            const wrapper = shallow(<HeroesContainer />);
            expect(wrapper).toMatchSnapshot();
        });

        test('renders an error message', (done) => {
            // Simulate an error
            fetchHeros.mockReturnValue(Promise.reject(new Error('Testing an error')));

            const wrapper = shallow(<HeroesContainer />);

            process.nextTick(() => {
                wrapper.update();
                expect(wrapper).toMatchSnapshot();
                done();
            });
        });

        test('renders heroes', (done) => {
            // Mock the response from the API
            fetchHeros.mockReturnValue(Promise.resolve(successfulResultMock));

            const wrapper = shallow(<HeroesContainer />);

            process.nextTick(() => {
                wrapper.update();
                expect(wrapper).toMatchSnapshot();
                wrapper.instance().componentWillUnmount();
                done();
            });
        });
    });

    describe('componentDidUpdate', () => {
        test('calls "fetchMarvelHeroes"', () => {
            const wrapper = shallow(<HeroesContainer />);
            const fetchMarvelHeroesSpy = jest.spyOn(wrapper.instance(), 'fetchMarvelHeroes');
            wrapper.setState({
                query: {
                    search: 'Test',
                    offset: 10,
                },
            });
            expect(fetchMarvelHeroesSpy).toHaveBeenCalledTimes(1);
            expect(fetchMarvelHeroesSpy).toHaveBeenCalledWith(10, 'Test');
        });
    });

    describe('componentWillUnmount', () => {
        test('calls "cancel" on CancelablePromise object', () => {
            const wrapper = shallow(<HeroesContainer />);
            const spy = jest.spyOn(wrapper.instance().cancelableFetch, 'cancel');
            wrapper.unmount();
            expect(spy).toHaveBeenCalled();
        });
    });

    describe('getThumbnailUrl', () => {
        test('calls "getImageUrl"', () => {
            const wrapper = shallow(<HeroesContainer />);
            const testHero = {};
            wrapper.instance().getThumbnailUrl(testHero);
            expect(getImageUrl).toHaveBeenCalledWith(testHero, 'small');
        });
    });

    describe('handleHeroClick', () => {
        test('renders the Modal with HeroDetails', (done) => {
            // Mock the response from the API
            fetchHeros.mockReturnValue(Promise.resolve(successfulResultMock));
            getImageUrl.mockReturnValue('getImageUrl result');

            const wrapper = shallow(<HeroesContainer />);

            process.nextTick(() => {
                wrapper.instance().handleHeroClick({
                    id: 1011334,
                    name: '3-D Man',
                    description: '',
                    modified: '2014-04-29T14:18:17-0400',
                    thumbnail: {
                        path: 'http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784',
                        extension: 'jpg',
                    },
                });
                expect(wrapper).toMatchSnapshot();
                done();
            });
        });
    });

    describe('handleCloseModal', () => {
        test('changes state', () => {
            const wrapper = shallow(<HeroesContainer />);
            wrapper.setState({
                selection: {
                    hero: {},
                },
            });
            expect(wrapper.state('selection')).not.toBeNull();
            wrapper.instance().handleCloseModal();
            expect(wrapper.state('selection')).toBeNull();
        });
    });

    describe('fetchHeroesPage', () => {
        let wrapper;
        beforeEach(() => {
            wrapper = shallow(<HeroesContainer pageSize="10" />);
        });

        test('calls setState setting "offset"', () => {
            wrapper.instance().fetchHeroesPage(3);
            expect(wrapper.state('query')).toEqual({
                offset: 20,
            });
        });

        test('calls setState keeping "search" state', () => {
            wrapper.setState({
                query: {
                    search: 'Test',
                },
            });
            wrapper.instance().fetchHeroesPage(4);
            expect(wrapper.state('query')).toEqual({
                offset: 30,
                search: 'Test',
            });
        });
    });

    describe('handleSearchChanged', () => {
        test('calls setState', () => {
            const wrapper = shallow(<HeroesContainer />);
            wrapper.instance().handleSearchChanged('Testing');
            expect(wrapper.state('query')).toEqual({
                search: 'Testing',
                offset: 0,
            });
        });
    });
});
