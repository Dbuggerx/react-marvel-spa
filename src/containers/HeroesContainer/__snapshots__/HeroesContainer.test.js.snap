// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroesContainer handleHeroClick renders the Modal with HeroDetails 1`] = `
<Fragment>
  <Modal
    onClose={[Function]}
    title="3-D Man"
  >
    <HeroDetails
      hero={
        Object {
          "description": "",
          "id": 1011334,
          "modified": "2014-04-29T14:18:17-0400",
          "name": "3-D Man",
          "thumbnail": Object {
            "extension": "jpg",
            "path": "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
          },
        }
      }
      imageUrl="getImageUrl result"
    />
  </Modal>
  <MainLayout
    info={
      <Information
        attribution={undefined}
      />
    }
    list={
      <HeroList
        getThumbnailUrl={[Function]}
        heroes={
          Array [
            Object {
              "description": "",
              "id": 1011334,
              "modified": "2014-04-29T14:18:17-0400",
              "name": "3-D Man",
              "thumbnail": Object {
                "extension": "jpg",
                "path": "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
              },
            },
            Object {
              "description": "Rick Jones has been Hulk's best bud since day one, but now he's more than a friend...he's a teammate! Transformed by a Gamma energy explosion, A-Bomb's thick, armored skin is just as strong and powerful as it is blue. And when he curls into action, he uses it like a giant bowling ball of destruction! ",
              "id": 1017100,
              "modified": "2013-09-18T15:54:04-0400",
              "name": "A-Bomb (HAS)",
              "thumbnail": Object {
                "extension": "jpg",
                "path": "http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16",
              },
            },
          ]
        }
        onHeroClick={[Function]}
      />
    }
    pagination={
      <PaginationContainer
        count={444}
        goToPage={[Function]}
        limit={222}
        offset={111}
        total={333}
      />
    }
    search={
      <SearchContainer
        lastSearch=""
        onSearchChanged={[Function]}
      />
    }
  />
</Fragment>
`;

exports[`HeroesContainer render renders a loading message 1`] = `<LoadingIndicator />`;

exports[`HeroesContainer render renders an error message 1`] = `
<ErrorMessage
  message="Error loading heroes"
/>
`;

exports[`HeroesContainer render renders heroes 1`] = `
<Fragment>
  <MainLayout
    info={
      <Information
        attribution={undefined}
      />
    }
    list={
      <HeroList
        getThumbnailUrl={[Function]}
        heroes={
          Array [
            Object {
              "description": "",
              "id": 1011334,
              "modified": "2014-04-29T14:18:17-0400",
              "name": "3-D Man",
              "thumbnail": Object {
                "extension": "jpg",
                "path": "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
              },
            },
            Object {
              "description": "Rick Jones has been Hulk's best bud since day one, but now he's more than a friend...he's a teammate! Transformed by a Gamma energy explosion, A-Bomb's thick, armored skin is just as strong and powerful as it is blue. And when he curls into action, he uses it like a giant bowling ball of destruction! ",
              "id": 1017100,
              "modified": "2013-09-18T15:54:04-0400",
              "name": "A-Bomb (HAS)",
              "thumbnail": Object {
                "extension": "jpg",
                "path": "http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16",
              },
            },
          ]
        }
        onHeroClick={[Function]}
      />
    }
    pagination={
      <PaginationContainer
        count={444}
        goToPage={[Function]}
        limit={222}
        offset={111}
        total={333}
      />
    }
    search={
      <SearchContainer
        lastSearch=""
        onSearchChanged={[Function]}
      />
    }
  />
</Fragment>
`;
